src/app/qgsmaptoolreshape.cpp

```diff
@@ -100,6 +100,28 @@ void QgsMapToolReshape::cadCanvasReleaseEvent( QgsMapMouseEvent * e )
         reshapeReturn = geom->reshapeGeometry( points() );
         if ( reshapeReturn == 0 )
         {
+          //avoid intersections on polygon layers
+          if ( vlayer->geometryType() == QGis::Polygon )
+          {
+
+            //ignore all current layer features as they should be reshaped too
+            QMap<QgsVectorLayer*, QSet<QgsFeatureId>> ignoreFeatures;
+            ignoreFeatures.insert( vlayer, vlayer->allFeatureIds() );
+
+            int avoidIntersectionsReturn = geom->avoidIntersections( ignoreFeatures );
+            if ( avoidIntersectionsReturn == 1 )
+            {
+              //not a polygon type. Impossible to get there
+            }
+            if ( geom->isGeosEmpty() ) //avoid intersection might have removed the whole geometry
```

Comment:
If avoidIntersections removes the whole geometry (e.g. the feature is fully inside another feature)